File: CVC_Tianming/codes/hyperspectral_image_processing/band_selection/MNBS.m
function IND = MNBS(Y,YN,n)
% Implementation of "A New Band Selection Method for Hyperspectral Image 
% Based on Data Quality"

S = Y-YN;
Cov_S = S.'*S;

Cov_N = YN.'*YN;

N = size(S,2);

k = N;

IND = 1:N; % index for selected bands

while k > n
    
    Q = zeros(1,k);
    
    for i = 1:k
        
        ind = setdiff(IND,IND(i));
        
        TEMP_S = Cov_S(ind,ind);
        [~,TEMP] = eig(TEMP_S);
        d_S = diag(TEMP);
        
        TEMP_N = Cov_N(ind,ind);
        [~,TEMP] = eig(TEMP_N);
        d_N = diag(TEMP);
        
        ratio = sum(log10(d_S./d_N));
        
        Q(i) = ratio;
        
    end
    
    [~,idx] = max(Q);
    
    fprintf('Among the %4d bands, removed band is %4d \n',k,IND(idx))
    
    IND = setdiff(IND,IND(idx)); % the one with low snr and high correlation is removed
    
    k = k-1;
    
end

File: hmi_fusion/models/hip/band_selection/MNBS.py
import numpy as np


def MNBS(Y, YN,n):
    #An Implementation of "A New Band Selection Method for Hyperspectral Image
    # Based on Data Quality"

    ##NOTE FROM CHASE: YN is your best guess on what the noise looks like
    # Independent Component analysis is a good way to get this guess

    S = Y - YN
    Cov_S = S.T @ S
    Cov_N = YN.T @ YN
    N = S.shape[1]
    k = N
    IND = list(range(N))

    while k > n:
        Q = np.zeros(k)

        for i in range(k):

            Temp_S = np.delete(arr=Cov_S,obj=IND[i],axis=0)
            Temp_S = np.delete(arr=Temp_S, obj=IND[i], axis=1)

            # _, TEMP = np.linalg.eig(Temp_S)
            #
            # d_S = np.diag(TEMP)

            TEMP = np.cov(Temp_S)

            d_S = np.linalg.det(TEMP)

            Temp_N = np.delete(arr=Cov_N,obj=IND[i],axis=0)
            Temp_N = np.delete(arr=Temp_N, obj=IND[i], axis=1)

            # _, TEMP = np.linalg.eig(Temp_N)
            #
            # d_N = np.diag(TEMP)

            TEMP = np.cov(Temp_N)

            d_N = np.linalg.det(TEMP)

            ratio = np.sum(np.log10(d_S/d_N))

            Q[i] = ratio

        idx = np.argmax(Q)

        print('Among the {} bands, removed band is {}'.format(k,IND[idx]))

        IND = np.delete(arr=IND, obj=idx,axis=0)

        k = k-1

    return IND